### ISSUE TITLE
System has not reached graphical.target

### ISSUE THREAD

**Issue by c4rlo** [Vladimir-csp/uwsm#28]
I'm trying to get the [Shell profile integration](https://github.com/Vladimir-csp/uwsm?tab=readme-ov-file#shell-profile-integration) to work. My `.bash_profile` snippet looks like this:

```sh
if uwsm check may-start -v; then
    exec uwsm start -S sway
fi
```

But instead of sway starting, I'm getting error message "System has not reached graphical.target" – same as in #27.

A few seconds later, I see the journal message saying that graphical.target has been reached.

If I then Ctrl+D out of the bash prompt, it logs me in again, and this time the sway launch works successfully.

**Comment by c4rlo** [comment]
To add a bit more detail:

- Arch Linux
- uwsm 0.15.7
- sway 1.9
- systemd 256.1
- dbus-broker 36
- Issue is consistently reproducible

**Comment by Vladimir-csp** [comment]
Does it happen only right after boot?

**Comment by c4rlo** [comment]
It happens after a fresh boot, as well as after `systemctl soft-reboot`. It does not happen when exiting sway; in that case, the auto-login on VT1 kicks in, `.bash_profile` gets re-executed, and uwsm/sway starts successfully immediately.

I'm wondering how this is normally supposed to work – what is supposed to cause `graphical.target` to have been reached by the time of the initial user login?

**Comment by Vladimir-csp** [comment]
Startup is async, `graphical.target` is probably linked as `default.target`, so it is the end goal. But `getty.target` does not depend on it and is reached earlier. Autologin just happens too fast.
I'm assuming autologin is configured via a drop-in for `getty@tty1.service` already, so try also adding this to it:
```
[Unit]
After=graphical.target
```

**Comment by c4rlo** [comment]
Thanks – that fixes it!

I'm actually a bit surprised that works, as I think I've now got a circular temporal dependency:

```
$ systemctl list-dependencies --before getty@tty1
getty@tty1.service
○ ├─rescue.service
○ ├─getty.target
● │ ├─multi-user.target
● │ │ ├─graphical.target
○ │ │ │ ├─shutdown.target
○ │ │ │   ├─systemd-soft-reboot.service
○ │ │ │   └─final.target
○ │ │ │     └─systemd-soft-reboot.service
  │ │ │ └─...
```

That final `...` I think indicates the circular dependency. The chain is `getty@tty1.service` → `getty.target` → `multi-user.target` → `graphical.target` → `getty@tty1.service`. But I guess systemd has some way of handling this that seems to work here (or I'm missing something)…

But I'm also wondering if wanting to sequence `uwsm start` after `graphical.target` is even necessary or desirable? If anything, I could imagine it making some sense the other way round – `graphical.target` coming _after_ the compositor being started.

**Comment by Vladimir-csp** [comment]
`graphical.target` is a system unit, display manager units wanted by it. So for example having `default.target` linked to `multi-user.target` instead will leave system in CLI mode.
Checking for `graphical.target` in `uwsm check may-start` serves as a way to replicate this behavior. Maybe this is a bad practice, I'll think about it.

**Comment by c4rlo** [comment]
I see. Yes that doesn't seem quite like the right way to do it; `uwsm check may-start` should probably (also?) check whether `graphical.target` is _intended_ to be reached as part of the current systemd transaction, not (just) whether it is already active.

How about something like:

- Create system unit `aiming-for-graphical.target` that has `Before=basic.target`, `WantedBy=graphical.target`, `DefaultDependencies=no`.
- Change `uwsm check may-start` to check whether `aiming-for-graphical.target` is active, instead of whether `graphical.target` is active.

**Comment by Vladimir-csp** [comment]
Intent of starting a unit is visible in systemd job listing. I'm currently a way to leverage this.

**Comment by c4rlo** [comment]
For reference, here is systemd commenting on the cycle introduced with `getty@tty.target` having `After=graphical.target`:

```
Fri 2024-06-28 20:55:02 BST systemd[1]: multi-user.target: Found ordering cycle on getty.target/start
Fri 2024-06-28 20:55:02 BST systemd[1]: multi-user.target: Found dependency on getty@tty1.service/start
Fri 2024-06-28 20:55:02 BST systemd[1]: multi-user.target: Found dependency on graphical.target/start
Fri 2024-06-28 20:55:02 BST systemd[1]: multi-user.target: Found dependency on multi-user.target/start
Fri 2024-06-28 20:55:02 BST systemd[1]: multi-user.target: Job getty.target/start deleted to break ordering cycle starting with multi-user.target/start
```

**Comment by Vladimir-csp** [comment]
I've got good results with jobs queue, solution is in progress, among some other things.

**Comment by novaluke** [comment]
A delayed `graphical.target` still causes problems - I'm getting a 5-second delay when logging in because [`uwsm start` still only checks for `graphical.target` to be already active, not activat***ing***](https://github.com/Vladimir-csp/uwsm/blob/eb61a15481c5c42860edd8929986d0f0413d6e84/uwsm/main.py#L4555).

In my case, it can take up to 60 seconds to reach `graphical.target` after reaching the login screen because I have services that start at boot but wait for the Tailscale interface to be up so they can bind to it, which can take a while. Since they start at boot, they have `WantedBy=multi-user.target`, which is wanted by `graphical.target` and thus `graphical.target` isn't considered `active` until they are also active.

The system is still perfectly usable in the meantime, however, so there should be no reason to prevent (or delay) logging in and starting Wayland. Perhaps some of the fix for `uwsm check` in ad397b26c8f4d8183b32c894b10ea40c9869d16f could be pulled in to `uwsm start`? Or is there some way to prevent these services from delaying `graphical.target` that I need to use?
